src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1512,9 +1519,21 @@
                                 :latest latest}
                        _ (when latest ;; TODO should we exclude plans here? would need to do it by latest boolean
                            ;; Mark existing latest report as historical
-                           ;; TODO what if this needs to create a partition?
-                           (jdbc/update! "resource_events" {:latest false} ["latest = true and report_id = ?" latest-report-id])
-                           (jdbc/update! "reports" {:latest false} ["latest = true and certname = ?" certname]))
+                           ;; If the partition does not exist throw ex-info to differentiate from a partition missing
+                           ;; from the case where the incoming report needs partition(s) made for it.
+                           (try
+                             (jdbc/update! "resource_events" {:latest false} ["latest = true and report_id = ?" latest-report-id])
+                             (jdbc/update! "reports" {:latest false} ["latest = true and certname = ?" certname])
+                             (catch java.sql.BatchUpdateException e
+                               ;; * 42P01 undefined table -- inheritance based partitions, or
+                               ;; declarative partitions when no children yet attached
+                               ;; * 23514 check constraint violation -- declarative partitions when row
+                               ;; does not match check constraint of existing children partitions
+                               (if (or (= "42P01" (.getSQLState e)) (= "23514" (.getSQLState e)))
+                                 (throw (ex-info "No partition to move previous latest report into"
+                                                 {:kind ::report-move-no-partition}
+                                                 e))
```

Comment:
Could use jdbc/sql-state and a `(#{:undefined-table :check-violation} (.getSqlState e))` or something.